File: soap-multilenguaje-main/soap-multilenguaje-main/ruby/soap-cliente/soap_cliente.rb
require 'net/http'
require 'uri'
require 'resolv'

def numero_a_letras_ingles(numero)
  url = URI("https://www.dataaccess.com/webservicesserver/NumberConversion.wso")

  # Forzamos resolucion IPv4 para evitar timeout por IPv6
  ipv4 = Resolv.getaddresses(url.host).find { |ip| ip =~ Resolv::IPv4::Regex }

  soap_envelope = <<~XML
    <?xml version="1.0" encoding="utf-8"?>
    <soap:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                   xmlns:xsd="http://www.w3.org/2001/XMLSchema"
                   xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
      <soap:Body>
        <NumberToWords xmlns="http://www.dataaccess.com/webservicesserver/">
          <ubiNum>#{numero}</ubiNum>
        </NumberToWords>
      </soap:Body>
    </soap:Envelope>
  XML

  http = Net::HTTP.new(url.host, url.port)
  http.ipaddr = ipv4 if ipv4
  http.use_ssl = true
  http.open_timeout = 10
  http.read_timeout = 10

  request = Net::HTTP::Post.new(url)
  request['Content-Type'] = 'text/xml; charset=utf-8'
  request['SOAPAction'] = 'http://www.dataaccess.com/webservicesserver/NumberToWords'
  request.body = soap_envelope

  response = http.request(request)
  match = response.body.match(%r{<m:NumberToWordsResult>(.*?)</m:NumberToWordsResult>})
  match ? match[1] : "No se pudo obtener el resultado"
end

if ARGV.empty?
  puts "Uso: ruby soap_cliente.rb <numero>"
  exit
end

numero = ARGV[0].to_i
letras_ingles = numero_a_letras_ingles(numero)
puts "Numero: #{numero}"
puts "En letras (ingles): #{letras_ingles}"

File: sopa-multilenguaje-main/perl/soap-cliente/soap_cliente.pl
use strict;
use warnings;
use LWP::UserAgent;

sub numero_a_letras_ingles {
    my ($numero) = @_;
    my $url = "https://www.dataaccess.com/webservicesserver/NumberConversion.wso";

    my $soap_envelope = qq{<?xml version="1.0" encoding="utf-8"?>
<soap:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
               xmlns:xsd="http://www.w3.org/2001/XMLSchema"
               xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
  <soap:Body>
    <NumberToWords xmlns="http://www.dataaccess.com/webservicesserver/">
      <ubiNum>$numero</ubiNum>
    </NumberToWords>
  </soap:Body>
</soap:Envelope>};

    my $ua = LWP::UserAgent->new(timeout => 15);
    my $response = $ua->post(
        $url,
        'Content-Type' => 'text/xml; charset=utf-8',
        'SOAPAction'   => 'http://www.dataaccess.com/webservicesserver/NumberToWords',
        Content        => $soap_envelope
    );

    my $body = $response->decoded_content;
    if ($body =~ m{<m:NumberToWordsResult>(.*?)</m:NumberToWordsResult>}) {
        return $1;
    }
    return "No se pudo obtener el resultado";
}

my $numero = $ARGV[0];
unless (defined $numero) {
    print "Uso: perl soap_cliente.pl <numero>\n";
    exit;
}

my $letras_ingles = numero_a_letras_ingles($numero);
print "Numero: $numero\n";
print "En letras (ingles): $letras_ingles\n";